pslab/bus/busio.py

```diff
@@ -0,0 +1,187 @@
+"""Circuitpython's busio compatibility layer for pslab-python.
+
+This module emulates the CircuitPython's busio API for devices or hosts running
+CPython or MicroPython using pslab-python.
+This helps to access many sensors using Adafruit's drivers via PSLab board.
+
+Notes
+-----
+Documentation:
+    https://circuitpython.readthedocs.io/en/6.3.x/shared-bindings/busio/index.html
+
+Examples
+--------
+Get humidity from Si7021 temperature and humidity sensor using adafruit_si7021 module.
+
+>>> import adafruit_si7021
+>>> from pslab.bus import busio   # import board, busio
+>>> i2c = busio.I2C()   # i2c = busio.I2C(board.SCL, board.SDA)
+>>> sensor = adafruit_si7021.SI7021(i2c)
+>>> print(sensor.relative_humidity)
+
+Get gyro reading from BNO055 using adafruit_bno055, board(just a wrapper for busio).
+
+>>> import adafruit_bno055
+>>> from pslab.bus import busio   # import board
+>>> i2c = busio.I2C()   # i2c = board.I2C()
+>>> sensor = adafruit_bno055.BNO055_I2C(i2c)
+>>> print(sensor.gyro)
+"""
+from typing import List, Union
+
+from pslab.bus.i2c import I2CPrimitive, I2CSlave
+from pslab.serial_handler import SerialHandler
+
+__all__ = "I2C"
+ReadableBuffer = Union[bytes, bytearray, memoryview]
+WriteableBuffer = Union[bytearray, memoryview]
+
+
+class I2C(I2CPrimitive):
+    """Busio I2C Class for CircuitPython Compatibility.
+
+    Parameters
+    ----------
+    device : :class:`SerialHandler`, optional
+        Serial connection to PSLab device. If not provided, a new one will be
+        created.
+    frequency : float, optional
+        Frequency of SCL in Hz.
+    """
+
+    def __init__(self, device: SerialHandler = None, frequency: int = 125e3):
+        # 125 kHz is as low as the PSLab can go.
+        super().__init__(device)
+        self._init()
+        self._configure(frequency)
+
+    def deinit(self) -> None:
+        """Just a dummy method."""
+        pass
+
+    def __enter__(self):
+        """Just a dummy method."""
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb) -> None:
+        """Just a dummy method."""
+        self.deinit()
```

Comment:
I wouldn't call these dummy methods, but since they aren't public they don't actually need docstrings.

Reply:
pydocstyle considers dunder method as public.

```
pslab/bus/busio.py:62 in public method `__enter__`:
        D105: Missing docstring in magic method
```

Reply:
Huh, TIL. Just refer to these as context managers in the docstrings, then.